### PR TITLE
expression: function SUBSTR/LEFT/RIGHT should manipulate bytes when parameter is a binary

### PR DESCRIPTION
**Pull request by zz-jason**
to https://github.com/pingcap/tidb/issues/4019

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @hanfei1991 PTAL

File: expression/builtin_string.go [hunk 1]
```diff
@@ -334,27 +339,59 @@ func (c *leftFunctionClass) getFunction(args []Expression, ctx context.Context)
 		return nil, errors.Trace(err)
 	}
 	bf.tp.Flen = args[0].GetType().Flen
+	if mysql.HasBinaryFlag(args[0].GetType().Flag) {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: 1. we may use setBinFlagOrBinStr in getFunction.
2. we should not use binary flag to check whether a string is a binary string.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -334,27 +339,59 @@ func (c *leftFunctionClass) getFunction(args []Expression, ctx context.Context)
...
+		types.SetBinChsClnFlag(bf.tp)
+		sig := &builtinLeftBinarySig{baseStringBuiltinFunc{bf}}
+		return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: move c.verifyArgs(args) to the first of this function.

**Comment by zz-jason** [comment]
@XuHuaiyu updated, PTAL again

**Comment by zz-jason** [comment]
@XuHuaiyu @zimulala @winkyao PTAL

File: expression/builtin_string.go [hunk 3]
```diff
@@ -329,32 +334,68 @@ type leftFunctionClass struct {
...
+	}
+	left, isNull, err := b.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	length := len(str)
+	l := int(left)
+	if l < 0 {
+		l = 0
+	} else if l > length {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: will this be ok?
``` go
else if length := len(str); l > length {
    l = length
}
```

File: expression/builtin_string.go [hunk 4]
```diff
@@ -329,32 +334,68 @@ type leftFunctionClass struct {
...
 	left, isNull, err := b.args[1].EvalInt(row, sc)
 	if isNull || err != nil {
-		return d, isNull, errors.Trace(err)
+		return "", true, errors.Trace(err)
 	}
-	runes := []rune(d)
+	runes := []rune(str)
 	length := len(runes)
 	l := int(left)
 	if l < 0 {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go [hunk 5]
```diff
@@ -371,33 +412,68 @@ type rightFunctionClass struct {
...
+	}
+	right, isNull, err := b.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	r := int(right)
+	length := len(str)
+	if r < 0 {
+		r = 0
+	} else if r > length {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go [hunk 6]
```diff
@@ -371,33 +412,68 @@ type rightFunctionClass struct {
...
 	right, isNull, err := b.args[1].EvalInt(row, sc)
 	if isNull || err != nil {
-		return d, isNull, errors.Trace(err)
+		return "", true, errors.Trace(err)
 	}
 	r := int(right)
-	runes := []rune(d)
+	runes := []rune(str)
 	length := len(runes)
 	if r < 0 {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go [hunk 7]
```diff
@@ -371,33 +412,68 @@ type rightFunctionClass struct {
...
 	if isNull || err != nil {
-		return d, isNull, errors.Trace(err)
+		return "", true, errors.Trace(err)
 	}
 	right, isNull, err := b.args[1].EvalInt(row, sc)
 	if isNull || err != nil {
-		return d, isNull, errors.Trace(err)
+		return "", true, errors.Trace(err)
 	}
 	r := int(right)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: r, runes := int(right), []rune(str)

File: expression/builtin_string.go [hunk 8]
```diff
@@ -329,32 +334,68 @@ type leftFunctionClass struct {
...
-		return d, isNull, errors.Trace(err)
+		return "", true, errors.Trace(err)
 	}
 	left, isNull, err := b.args[1].EvalInt(row, sc)
 	if isNull || err != nil {
-		return d, isNull, errors.Trace(err)
+		return "", true, errors.Trace(err)
 	}
-	runes := []rune(d)
+	runes := []rune(str)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: runes, l := []rune(str), int(left)

File: expression/builtin_string.go [hunk 9]
```diff
@@ -744,46 +820,82 @@ type substringFunctionClass struct {
...
-
-	hasLen := len(args) == 3
-	if hasLen {
-		bf, err = newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt, tpInt)
-	} else {
-		bf, err = newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt)
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(args) == 3 {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: this code is really ugly, too many duplicate codes.

will this be better..
``` go
switch len(args) {
case 2:
   bf, err = newBaseBuiltinFuncWithTp()
case 3:
   bf, err = newBaseBuiltinFuncWithTp()
}
checkErr
argType = xxx
isBinary := xxx
switch {
   case isBinaryStr && len(args) == 3:
   case 
   case 
   case
}
return sig.setSelf(sig)
```

**Comment by zz-jason** [comment]
@XuHuaiyu updated, PTAL again.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by zz-jason** [comment]
@breeswish @jackysp @zimulala @tiancaiamao PTAL

File: expression/builtin_string.go [hunk 10]
```diff
@@ -329,82 +334,146 @@ type leftFunctionClass struct {
...
 	}
-	runes := []rune(d)
-	length := len(runes)
-	l := int(left)
-	if l < 0 {
-		l = 0
-	} else if l > length {
-		l = length
+	leftLength := int(left)
+	if strLength := len(str); leftLength > strLength {
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: `if leftLength > len(str) {`

File: expression/builtin_string.go [hunk 11]
```diff
@@ -329,82 +334,146 @@ type leftFunctionClass struct {
...
+	str, isNull, err := b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	left, isNull, err := b.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	runes, leftLength := []rune(str), int(left)
+	if runeLength := len(runes); leftLength > runeLength {
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go [hunk 12]
```diff
@@ -744,46 +813,82 @@ type substringFunctionClass struct {
...
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	pos, isNull, err := b.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	length := int64(len(str))
+	if pos < 0 {
+		pos += length
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: "string"  -1 get "strin" ?
"string"  -1111, what's the result, "string"?

File: expression/builtin_string.go [hunk 13]
```diff
@@ -329,82 +334,146 @@ type leftFunctionClass struct {
...
-	l := int(left)
-	if l < 0 {
-		l = 0
-	} else if l > length {
-		l = length
+	leftLength := int(left)
+	if strLength := len(str); leftLength > strLength {
+		leftLength = strLength
+	} else if leftLength < 0 {
+		leftLength = 0
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: Is this mysql behavior?

`select left("测试", "-1")` get empty string?

File: expression/integration_test.go [hunk 14]
```diff
@@ -413,6 +413,13 @@ func (s *testIntegrationSuite) TestStringBuiltin(c *C) {
 	result.Check(testkit.Rows(" de e"))
 	result = tk.MustQuery(`select right("中文abc", 2), right("中文abc", 4), right("中文abc", 5)`)
 	result.Check(testkit.Rows("bc 文abc 中文abc"))
+	tk.MustExec("drop table if exists t")
```
Comment (tiancaiamao) [inline expression/integration_test.go]: Is there any test position is negative?

File: expression/builtin_string.go (hunk 13 above)
Comment (zz-jason) [inline expression/builtin_string.go]: yse

File: expression/builtin_string.go (hunk 10 above)
Comment (zz-jason) [inline expression/builtin_string.go]: if we do so, `len(str)` will be called again in line 373. prefer to retain

File: expression/builtin_string.go (hunk 11 above)
Comment (zz-jason) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go (hunk 12 above)
Comment (zz-jason) [inline expression/builtin_string.go]: - `select substr("string", -1);` returns "g"
- `select substr("string", -1111);` returns an empty string.

result of TiDB and MySQL is the same.

File: expression/integration_test.go (hunk 14 above)
Comment (zz-jason) [inline expression/integration_test.go]: added.

**Comment by zz-jason** [comment]
@tiancaiamao updated, PTAL again.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**